src/crimson/osd/heartbeat.cc

```diff
@@ -98,6 +98,87 @@ void Heartbeat::add_peer(osd_id_t peer, epoch_t epoch)
   }
 }
 
+seastar::future<Heartbeat::osds_t> Heartbeat::remove_down_peers()
+{
+  osds_t osds;
+  for (auto& peer : peers) {
+    osds.push_back(peer.first);
+  }
```

Comment:
could use `boost::copy` + `boost::adaptors::map_keys`, but i think that'd slow down the compiling .